### ISSUE TITLE
[firebase_messaging] No documentation for Android v2 embedding backgroundMessage handler

### ISSUE THREAD

**Issue by VictorUvarov** [FirebaseExtended/flutterfire#1775]
**Describe the bug**
Currently the documentation does **NOT** support the new Flutter Android v2 embedding that came along with flutter 1.12.13. The function `GeneratedPluginRegistrant.registerWith(registry);` no longer accepts a `io.flutter.plugin.common.PluginRegistry` object, but a `io.flutter.embedding.engine.FlutterEngine` object.

Current code in documentation:
```java
 import io.flutter.app.FlutterApplication;
 import io.flutter.plugin.common.PluginRegistry;
 import io.flutter.plugin.common.PluginRegistry.PluginRegistrantCallback;
 import io.flutter.plugins.GeneratedPluginRegistrant;
 import io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService;

 public class Application extends FlutterApplication implements PluginRegistrantCallback {
   @Override
   public void onCreate() {
     super.onCreate();
     FlutterFirebaseMessagingService.setPluginRegistrant(this);
   }

   @Override
   public void registerWith(PluginRegistry registry) {
     GeneratedPluginRegistrant.registerWith(registry);
   }
 }
```

Newer code attempt with error:
```java
import io.flutter.app.FlutterApplication;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugins.GeneratedPluginRegistrant;
import io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService;

public class Application extends FlutterApplication implements PluginRegistry.PluginRegistrantCallback {
    @Override
    public void onCreate() {
        super.onCreate();
        FlutterFirebaseMessagingService.setPluginRegistrant(this);
    }

    @Override
    public void registerWith(PluginRegistry registry) {
        GeneratedPluginRegistrant.registerWith(registry); // ERROR: Expects io.flutter.embedding.engine.FlutterEngine
    }
}
```

**To Reproduce**
1. Run `flutter create -a java example`
2. Follow firebase_messaging background messaging instructions
4. Notice GeneratedPluginRegistrant.registerWith(registry) no longer expects a PluginRegistry type

**Comment by stoppiNeobiz** [comment]
any solution for this error?

**Comment by sgehrman** [comment]
same issue

**Comment by gakubhat** [comment]
Solved it with following:
`app/src/build.gradle`
add the following:
`    implementation "com.google.firebase:firebase-messaging:20.1.0"`
Application.kt
```kt
import io.flutter.app.FlutterApplication
import io.flutter.plugin.common.PluginRegistry
import io.flutter.plugin.common.PluginRegistry.PluginRegistrantCallback
import io.flutter.plugins.firebasemessaging.FlutterFirebaseMessagingService

class Application : FlutterApplication(), PluginRegistrantCallback {
    override fun onCreate() {
        super.onCreate()
        FlutterFirebaseMessagingService.setPluginRegistrant(this)
    }

    override fun registerWith(registry: PluginRegistry?) {
        registry?.registrarFor("io.flutter.plugins.firebasemessaging.FirebaseMessagingPlugin");
    }


}
```

**Comment by woprandi** [comment]
That would be better to get an official fix and not have to do a workaround with potential side effect

**Comment by woprandi** [comment]
I have another issue with the @gakubhat workaround (https://github.com/FirebaseExtended/flutterfire/issues/116)
I followed this workaround : https://www.djamware.com/post/5e4b26e26cdeb308204b427f/flutter-tutorial-firebase-cloud-messaging-fcm-push-notification

Frustrating to have so many issues on a important plugin...

**Comment by Zazo032** [comment]
@Ehesp hi, is there any ETA for solving this?

**Comment by FabulousGee** [comment]
This issue has been opened about half a year ago and still not under real investigation? C'mon guys, this is a real show stopper, especially for newbies that want to get into this stuff...

Just set up a new, standard flutter project and follow the current documentation. It won't work...

**Comment by kroikie** [comment]
@All Sorry about the delay here, we are currently working on updating the messaging plugin, the documentation will be updated according to those final updates. We should have more on this very soon.

**Comment by videjunior** [comment]
`class MainActivity : FlutterActivity(), PluginRegistry.PluginRegistrantCallback {
    private val CHANNEL_PLACES = "br.com.urbmobilidade.driver/places"

    private var placesPlugin: PlacesPlugin? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        placesPlugin = PlacesPlugin()
        createNotificationChannel()
        FlutterFirebaseMessagingService.setPluginRegistrant(this)
    }

    override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
        GeneratedPluginRegistrant.registerWith(flutterEngine)

        MethodChannel(flutterEngine.dartExecutor.binaryMessenger, CHANNEL_PLACES).setMethodCallHandler {
            call, result ->

            when(call.method) {
                "initInstance" -> placesPlugin?.initInstance(context, call.argument<String>("apiKey")!!, result)
                "getPredictions" -> placesPlugin?.getPredictions(call.argument<String>("query")!!, call.argument<HashMap<String, Double>>("origin")!!, call.argument<Double>("radius")!!, result)
                "getPlacesDetails" -> placesPlugin?.getPlacesDetails(call.argument<String>("placeId")!!, result)
                "getCurrentPlace" -> placesPlugin?.getCurrentPlace(result)
                else -> result.notImplemented()
            }
        }
    }

    override fun registerWith(registry: PluginRegistry?) {
        registry?.let { FirebaseCloudMessagingPluginRegistrant.registerWith(it) }
    }
}`

`class FirebaseCloudMessagingPluginRegistrant {
    companion object {
        fun registerWith(registry: PluginRegistry) {
            if (alreadyRegisteredWith(registry)) {
                return;
            }
            FirebaseMessagingPlugin.registerWith(registry.registrarFor("io.flutter.plugins.firebasemessaging.FirebaseMessagingPlugin"))
        }

        fun alreadyRegisteredWith(registry: PluginRegistry): Boolean {
            val key = FirebaseCloudMessagingPluginRegistrant::class.java.name
            if (registry.hasPlugin(key)) {
                return true
            }
            registry.registrarFor(key)
            return false
        }
    }
}`

Funcionou no meu projeto já modificado para o embedding v2.

**Comment by videjunior** [comment]
É só você usar o PluginRegistry.PluginRegistrantCallback dessa forma:

``` kt
class FirebaseCloudMessagingPluginRegistrant {
    companion object {
        fun registerWith(registry: PluginRegistry) {
            if (alreadyRegisteredWith(registry)) {
                return;
            }
            FirebaseMessagingPlugin.registerWith(registry.registrarFor("io.flutter.plugins.firebasemessaging.FirebaseMessagingPlugin"))
        }

        fun alreadyRegisteredWith(registry: PluginRegistry): Boolean {
            val key = FirebaseCloudMessagingPluginRegistrant::class.java.name
            if (registry.hasPlugin(key)) {
                return true
            }
            registry.registrarFor(key)
            return false
        }
    }
}
```
``` kt
class MainActivity : FlutterActivity(), PluginRegistry.PluginRegistrantCallback {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        FlutterFirebaseMessagingService.setPluginRegistrant(this)
    }

    override fun configureFlutterEngine(@NonNull flutterEngine: FlutterEngine) {
        GeneratedPluginRegistrant.registerWith(flutterEngine)
    }

    override fun registerWith(registry: PluginRegistry?) {
        registry?.let { FirebaseCloudMessagingPluginRegistrant.registerWith(it) }
    }
}
```
Funciona normalmente o "backgroundMessage handler" no meu projeto já modificado para embedding V2.

**Comment by josh-burton** [comment]
I'm a bit confused about which plugins need to be manually registered in the `registerWith` method.

Is it any plugin that we are using in the background, or is it plugins that haven't yet updated to the v2 embedding?

**Comment by pangj** [comment]
Has this issue be resolved now, anybody knows? It is really important plugin for the Flutter framework.

**Comment by lazylazyllama** [comment]
It will be resolved very very soon, when version 8.0.0 will be released.

**Comment by Salakar** [comment]
Hey all 👋 

As part of our roadmap (#2582) we've just shipped a complete rework of the `firebase_messaging` plugin that aims to solve this and many other issues. Specifically v2 embedding is now supported and requires no documentation/manual steps [unless you still need to use v1 embedding](https://firebase.flutter.dev/docs/messaging/overview#3-android-integration).

If you can, please try out the dev release (see the [migration guide](https://firebase.flutter.dev/docs/migration/) for upgrading and for changes) and if you have any feedback then join in the discussion [here](https://github.com/FirebaseExtended/flutterfire/discussions/4023).

Given the scope of the rework I'm going to go ahead and close this issue in favor of trying out the latest plugin.

Thanks everyone.
